### PR TITLE
feat(core): provide secrets to template strings

### PR DESCRIPTION
**Pull request by thsig**
For projects connected with Garden Cloud, any secrets for the project in the given environment can now be used in template strings.

### CODE REVIEW

File: examples/build-dependencies/backend/main.go [hunk 1]
```diff
@@ -3,10 +3,11 @@ package main
 import (
 	"fmt"
 	"net/http"
+	"os"
 )
 
 func handler(w http.ResponseWriter, r *http.Request) {
-	fmt.Fprint(w, "Hello from Go!")
+	fmt.Fprint(w, os.Getenv("THE_SEC"))
```
Comment (eysi09) [inline examples/build-dependencies/backend/main.go:10]: Should this be here?

File: examples/demo-project/backend/garden.yml [hunk 2]
```diff
@@ -4,6 +4,8 @@ description: Backend service container
 type: container
 services:
   - name: backend
+    env:
```
Comment (eysi09) [inline examples/demo-project/backend/garden.yml:7]: Same as above.

File: garden-service/src/secrets/gardencloud/getSecret.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+/*
```
Comment (eysi09) [inline garden-service/src/secrets/gardencloud/getSecret.ts:1]: File name is camel cased whereas we usually kebab case file names.

File: garden-service/package.json [hunk 4]
```diff
@@ -108,6 +108,7 @@
     "request": "^2.88.2",
     "request-promise": "^4.2.5",
     "slash": "^3.0.0",
+    "request-promise-native": "^1.0.8",
```
Comment (eysi09) [inline garden-service/package.json:111]: Is this used anywhere?

File: garden-service/src/garden.ts [hunk 5]
```diff
@@ -294,6 +299,13 @@ export class Garden {
         throw new InternalError(errMsg, {})
       } else {
         await login(platformUrl, log)
+        secrets = await getSecrets({
```
Comment (eysi09) [inline garden-service/src/garden.ts:302]: What happens if `secrets` don't contain all the secrets the config expects. Say you have something like:

```yaml
services:
# ...
  env:
    API_KEY: ${secrets.apiKey}
```

and `secrets.apiKey` is undefined. 

Does it just deploy the service with an empty `API_KEY`? Or is this handled somewhere?

Comment (solomonope) [reply garden-service/src/garden.ts:302]: I think this is a valid question, from a fail early point of view, I think garden should fail that secrets were not resolved. I think that is what is currently happening

Comment (thsig) [reply garden-service/src/garden.ts:308]: An error will be thrown if the key is not defined (see https://github.com/garden-io/garden/blob/secrets-in-template-strings/garden-service/test/unit/src/config/config-context.ts#L343).

This leans on our generic error handling for undefined keys in template strings.

We are not, however, validating that the value is truthy (so a secret with a `null`/`undefined` value for a valid/present key would not result in an error).

File: garden-service/src/cloud/secrets/garden-cloud/get-secret.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  try {
+    const url = `${platformUrl}/secrets/project/name/${projectConfig.name}/env/${environmentName}`
+    const headers = { "x-access-auth-token": clientAuthToken }
+    const res = await got(url, { headers }).json<GotResponse<any>>()
+    if (res && res["status"] === "success") {
+      return res["data"]
+    }
+    return {}
+  } catch (err) {
+    log.error("An error occurred while fetching secrets for the project.")
```
Comment (eysi09) [inline garden-service/src/cloud/secrets/garden-cloud/get-secret.ts:28]: Feels like this should be a clearer error message. Could this e.g. be an auth error? In which case we should instruct the user to log in.

File: garden-service/src/garden.ts (hunk 5 above)
Comment (eysi09) [reply garden-service/src/garden.ts:308]: All right. Wondering how we should in general handle platforrm related errors. For example if the secret is missing from the backend you want a better error than just: "Failed resolving template string..." (or whatever it is).

Comment (thsig) [reply garden-service/src/garden.ts:308]: I've added some error handling around empty secret values (in a TBS commit).

File: garden-service/src/cloud/secrets/garden-cloud/get-secret.ts (hunk 6 above)
Comment (thsig) [reply garden-service/src/cloud/secrets/garden-cloud/get-secret.ts:29]: Shall we address this in a separate PR?

File: garden-service/src/template-string.ts [hunk 7]
```diff
@@ -131,6 +131,10 @@ export function getModuleTemplateReferences<T extends object>(obj: T) {
   return refs.filter((ref) => ref[0] === "modules" && ref.length > 1)
 }
 
+export function detectMissingSecrets<T extends object>(obj: T, secrets: PrimitiveMap) {
```
Comment (eysi09) [inline garden-service/src/template-string.ts:134]: What's this for?

File: garden-service/src/garden.ts [hunk 8]
```diff
@@ -663,6 +663,10 @@ export class Garden {
 
     this.log.silly(`Resolving module configs`)
 
+    if (Object.keys(this.secrets).length > 0) {
+      // Gather secret references from project config and raw module configs
```
Comment (eysi09) [inline garden-service/src/garden.ts:667]: Something missing here?

File: garden-service/src/template-string.ts (hunk 7 above)
Comment (thsig) [reply garden-service/src/template-string.ts:134]: Ah, my bad I had some uncommited on a separate branch. Will update tomorrow.

**Pull request merged**